clang/lib/Sema/SemaBounds.cpp

```diff
@@ -3684,21 +3684,33 @@ namespace {
       Expr *SubExpr = E->getSubExpr()->IgnoreParens();
 
       if (isa<CompoundLiteralExpr>(SubExpr)) {
+        // The lvalue bounds of a struct-typed compound literal expression e
+        // are bounds(&value(temp(e), &value(temp(e)) + 1).
+        if (E->getType()->isStructureType()) {
+          Expr *TempUse = CreateTemporaryUse(E);
+          Expr *Addr = CreateAddressOfOperator(TempUse);
+          const llvm::APInt One(Context.getTargetInfo().getPointerWidth(0), 1);
+          IntegerLiteral *Size = CreateIntegerLiteral(One);
+          CountBoundsExpr *CBE = new (Context)
+              CountBoundsExpr(BoundsExpr::Kind::ElementCount, Size,
+                              SourceLocation(), SourceLocation());
+          return ExpandToRange(Addr, CBE);
+        }
+
         // Only expressions with array or function type can have a decayed
-        // type, which is used to create the lvalue bounds.  Compound literals
-        // with non-array, non-function types do not have lvalue bounds.
-        // TODO: checkedc-clang issue #870: bind all compound literals to
-        // temporaries and infer lvalue bounds for struct compound literals.
-        if (!(E->getType()->isArrayType() || E->getType()->isFunctionType()))
-          return CreateBoundsAlwaysUnknown();
+        // type, which is used to create the lvalue bounds.
+        if (E->getType()->isArrayType() || E->getType()->isFunctionType()) {
```

Comment:
I don't think function types should have bounds, so I'm confused by the prior code.  Should function types be given unknown bounds instead?